### ISSUE TITLE
Include transitive dependency jars

### ISSUE THREAD

**Issue by rouzwawi** [siom79/japicmp#65]
I have a multi module project with a module that is used from other modules. That common module in turn depends on Guice and exposes a class that extends AbstractModule.

```
guice
`- a
   `- b
   `- c
```

I have to declare guice as a dependency in the plugin configuration for both `b` and `c` or I'll get an error:

```
[ERROR] Failed to execute goal com.github.siom79.japicmp:japicmp-maven-plugin:0.5.1:cmp (default) on project apollo-nameless: Execution default of goal com.github.siom79.japicmp:japicmp-maven-plugin:0.5.1:cmp failed: Could not load 'com.example.ExampleModule': com.google.inject.AbstractModule. Please make sure that all libraries have been added to the classpath...
```

Is there a good reason for not including transitive dependencies automatically?

**Comment by siom79** [comment]
Hi rouzwawi,

I have tried to reconstruct your issue. The pom.xml of my base project looks like this one:
```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <modelVersion>4.0.0</modelVersion>

    <groupId>japicmp-test</groupId>
    <artifactId>japicmp-test-base</artifactId>
    <packaging>pom</packaging>
    <version>1.0-SNAPSHOT</version>
    <modules>
        <module>japicmp-test-commons</module>
    </modules>
    <dependencies>
        <dependency>
            <groupId>com.google.guava</groupId>
            <artifactId>guava</artifactId>
            <version>18.0</version>
        </dependency>
    </dependencies>
</project>
```
The child module `japicmp-test-commons` has the following pom.xml:
```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <parent>
        <artifactId>japicmp-test-base</artifactId>
        <groupId>japicmp-test</groupId>
        <version>1.0-SNAPSHOT</version>
    </parent>
    <modelVersion>4.0.0</modelVersion>

    <artifactId>japicmp-test-commons</artifactId>

    <build>
        <plugins>
            <plugin>
                <groupId>com.github.siom79.japicmp</groupId>
                <artifactId>japicmp-maven-plugin</artifactId>
                <version>0.5.1</version>
                <configuration>
                    <oldVersion>
                        <file>
                            <path>${project.build.directory}/${project.artifactId}-${project.version}.jar</path>
                        </file>
                    </oldVersion>
                    <newVersion>
                        <file>
                            <path>${project.build.directory}/${project.artifactId}-${project.version}.jar</path>
                        </file>
                    </newVersion>
                </configuration>
                <executions>
                    <execution>
                        <phase>verify</phase>
                        <goals>
                            <goal>cmp</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
        </plugins>
    </build>
</project>
```
Now the `japicmp` plugin outputs:
```
[INFO] --- japicmp-maven-plugin:0.5.1:cmp (default) @ japicmp-test-commons ---
[INFO] /opt/maven/repository/guava-18.0.jar; scope: compile
...
```
Note that the second `[INFO]` line denotes that the guava library from the parent pom has been added to the classpath. Hence I was not able to reproduce the issue.

Can you check with `mvn help:effective-pom` within the child module's directory that the guava dependency you are missing is shown in the `dependencies` section?

**Comment by rouzwawi** [comment]
I've tried to isolate the problem a bit more now and I have an example that triggers it here: https://github.com/rouzwawi/japicmp-test

The setup is two modules `a` and `b` where `a` depends on `guava` and exposes a type `SpecialFuture extends AbstractFuture`. `b` in turn depends on `a` and exposes `SpecialFuture` in it's public api.

I get:

```
[ERROR] Failed to execute goal com.github.siom79.japicmp:japicmp-maven-plugin:0.5.1:cmp (default) on project test-leaf: Execution default of goal com.github.siom79.japicmp:japicmp-maven-plugin:0.5.1:cmp failed: Could not load 'com.example.Using$MyFuture': com.google.common.util.concurrent.AbstractFuture. Please make sure that all libraries have been added to the classpath (CLASSPATH=/usr/local/Cellar/maven/3.3.3/libexec/boot/plexus-classworlds-2.5.2.jar:/Users/rouz/code/japicmp-transitive/a/target/test-common-1.0.0-SNAPSHOT.jar) or try the option '--ignore-missing-classes'. -> [Help 1]
```

While trying to provoke the error I noticed that it only happens when my jar somehow exposes a type from the transitive dependency. In this case, `MyFuture` extends `SpecialFuture` which extends `AbstractFuture` from guava.

The same thing happens if I have this in my public api

```java
// exposed somehwere in test-leaf
public SpecialFuture foo() {
	return new SpecialFuture() {};
}
```

**Comment by siom79** [comment]
Thank you for setting up such a detailed project that triggers the problem. With its help I could track down the issue in the japicmp maven plugin. The solution is rather simple (see [here](http://ars-codia.raphaelbauer.com/2014/01/maven-mojo-getting-dependencies-of.html)).

I have committed the fix on the `develop` branch. With your project setup this worked.

If you want you can try it yourself:
```
git clone https://github.com/siom79/japicmp.git
cd japicmp
git checkout develop
mvn clean install
```
Then update the version of the japicmp maven plugin in your sample project to `0.5.2-SNAPSHOT` and rebuild it.

**Comment by rouzwawi** [comment]
Neat fix! And i just tried the develop branch on my real project and it all works fine!

Thanks!

**Comment by adrienlauer** [comment]
I recently tried the japicmp maven plugin in my project and this fix is a must have for me. It's working perfectly with the develop branch version.

**Comment by rouzwawi** [comment]
@siom79 do you have an ETA on 0.5.2?

**Comment by siom79** [comment]
As soon as I have feedback for issue [#66](https://github.com/siom79/japicmp/issues/66), I will release these two issues.

**Comment by siom79** [comment]
Released with version 0.5.2.
